hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java

```diff
@@ -342,6 +343,13 @@ public Response put(
           output.getMetadata().put(ETAG, eTag);
         }
       }
+
+      String encodedClientETag = headers.getHeaderString(CHECKSUM_HEADER);
+      if (S3Utils.isEtagMisMatch(encodedClientETag, eTag)) {
+        delete(bucketName, keyPath, uploadID, null);
```

Comment:
I'm a bit worried about this since if there are concurrent key creation from the user, it might delete other user's key instead. I would prefer if the ETag checking is done before the key is committed (i.e. before `OzoneOutputStream` is closed). In the case of multipart uploads, user can decide to continue uploading the same part again, instead of aborting the multipart uploads (that might already contain a lot of large parts).

Since we have an open key cleanup service, the single / multipart upload open keys will be deleted eventually.